File: apps/api/app/db/models/reassessment.py
"""B14 reassessment instantiation and mastery delta projection (PEV-043)."""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import (
    CheckConstraint,
    DateTime,
    ForeignKey,
    Index,
    Integer,
    Numeric,
    String,
    UniqueConstraint,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, TimestampMixin, UUIDPrimaryKeyMixin

ALGORITHM_VERSION_B14_V1 = "B14_V1"


class Reassessment(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "reassessments"
    __table_args__ = (
        UniqueConstraint(
            "tenant_id",
            "improvement_assessment_id",
            name="uq_reassessments_tenant_blueprint",
        ),
        UniqueConstraint("assessment_id", name="uq_reassessments_assessment"),
        CheckConstraint(
            "status IN ('CREATED','SUBMITTED','PUBLISHED')",
            name="ck_reassessments_status",
        ),
        Index("ix_reassessments_tenant_student", "tenant_id", "student_id"),
        Index(
            "ix_reassessments_tenant_blueprint",
            "tenant_id",
            "improvement_assessment_id",
        ),
        Index("ix_reassessments_tenant_assessment", "tenant_id", "assessment_id"),
        Index("ix_reassessments_tenant_submission", "tenant_id", "submission_id"),
        Index(
            "ix_reassessments_tenant_published_result",
            "tenant_id",
            "published_result_id",
        ),
    )

    tenant_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("tenants.id", ondelete="CASCADE")
    )
    improvement_assessment_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("improvement_assessments.id", ondelete="RESTRICT")
    )
    student_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("students.id", ondelete="RESTRICT")
    )
    curriculum_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("curricula.id", ondelete="RESTRICT")
    )
    assessment_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("assessments.id", ondelete="RESTRICT")
    )
    assessment_version_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("assessment_versions.id", ondelete="RESTRICT")
    )
    submission_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("submissions.id", ondelete="SET NULL"), nullable=True
    )
    published_result_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("published_results.id", ondelete="SET NULL"), nullable=True
    )
    status: Mapped[str] = mapped_column(String(32))
    instantiation_hash: Mapped[str] = mapped_column(String(64))
    baseline_captured_at: Mapped[datetime] = mapped_column(DateTime(timezone=True))
    algorithm_version: Mapped[str] = mapped_column(
        String(32),
        default=ALGORITHM_VERSION_B14_V1,
        server_default=ALGORITHM_VERSION_B14_V1,
    )
    created_by: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"), nullable=True
    )


class ReassessmentItem(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "reassessment_items"
    __table_args__ = (
        UniqueConstraint(
            "reassessment_id",
            "improvement_assessment_item_id",
            name="uq_reassessment_items_blueprint_item",
        ),
        UniqueConstraint(
            "question_version_id",
            name="uq_reassessment_items_question_version",
        ),
        Index(
            "ix_reassessment_items_tenant_reassessment",
            "tenant_id",
            "reassessment_id",
        ),
    )

    tenant_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("tenants.id", ondelete="CASCADE")
    )
    reassessment_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("reassessments.id", ondelete="CASCADE")
    )
    improvement_assessment_item_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("improvement_assessment_items.id", ondelete="RESTRICT")
    )
    question_version_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("question_versions.id", ondelete="RESTRICT")
    )
    curriculum_node_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("curriculum_nodes.id", ondelete="RESTRICT")
    )
    item_code_snapshot: Mapped[str] = mapped_column(String(100))
    template_kind_snapshot: Mapped[str] = mapped_column(String(64))
    question_template_ref_snapshot: Mapped[str | None] = mapped_column(
        String(255), nullable=True
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )


class ReassessmentMasteryDelta(UUIDPrimaryKeyMixin, Base):
    __tablename__ = "reassessment_mastery_deltas"
    __table_args__ = (
        UniqueConstraint(
            "tenant_id",
            "reassessment_id",
            "curriculum_node_id",
            "algorithm_version",
            name="uq_reassessment_mastery_deltas_grain",
        ),
        Index(
            "ix_reassessment_mastery_deltas_tenant_reassessment",
            "tenant_id",
            "reassessment_id",
        ),
    )

    tenant_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("tenants.id", ondelete="CASCADE")
    )
    reassessment_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("reassessments.id", ondelete="CASCADE")
    )
    curriculum_node_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("curriculum_nodes.id", ondelete="RESTRICT")
    )
    baseline_concept_mastery: Mapped[Decimal | None] = mapped_column(
        Numeric(10, 6), nullable=True
    )
    baseline_execution_accuracy: Mapped[Decimal | None] = mapped_column(
        Numeric(10, 6), nullable=True
    )
    baseline_concept_decisive_count: Mapped[int] = mapped_column(Integer, default=0)
    baseline_execution_decisive_count: Mapped[int] = mapped_column(Integer, default=0)
    baseline_concept_inconclusive_count: Mapped[int] = mapped_column(
        Integer, default=0
    )
    baseline_execution_inconclusive_count: Mapped[int] = mapped_column(
        Integer, default=0
    )
    baseline_evidence_count: Mapped[int] = mapped_column(Integer, default=0)
    baseline_source_evidence_hash: Mapped[str] = mapped_column(String(64))
    post_snapshot_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("mastery_state_snapshots.id", ondelete="SET NULL"), nullable=True
    )
    post_published_result_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("published_results.id", ondelete="SET NULL"), nullable=True
    )
    post_concept_mastery: Mapped[Decimal | None] = mapped_column(
        Numeric(10, 6), nullable=True
    )
    post_execution_accuracy: Mapped[Decimal | None] = mapped_column(
        Numeric(10, 6), nullable=True
    )
    post_concept_decisive_count: Mapped[int | None] = mapped_column(
        Integer, nullable=True
    )
    post_execution_decisive_count: Mapped[int | None] = mapped_column(
        Integer, nullable=True
    )
    post_concept_inconclusive_count: Mapped[int | None] = mapped_column(
        Integer, nullable=True
    )
    post_execution_inconclusive_count: Mapped[int | None] = mapped_column(
        Integer, nullable=True
    )
    post_evidence_count: Mapped[int | None] = mapped_column(Integer, nullable=True)
    post_source_evidence_hash: Mapped[str | None] = mapped_column(
        String(64), nullable=True
    )
    concept_delta: Mapped[Decimal | None] = mapped_column(Numeric(10, 6), nullable=True)
    execution_delta: Mapped[Decimal | None] = mapped_column(
        Numeric(10, 6), nullable=True
    )
    materialized_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    algorithm_version: Mapped[str] = mapped_column(
        String(32),
        default=ALGORITHM_VERSION_B14_V1,
        server_default=ALGORITHM_VERSION_B14_V1,
    )

File: apps/web/src/lib/api/http/reassessment.ts
import type {
  B14RebuildResult,
  Reassessment,
  ReassessmentInstantiateRequest,
  ReassessmentItem,
  ReassessmentMasteryDelta,
} from "@/lib/types/domain";
import { httpRequest } from "./client";

function asNumber(
  value: number | string | null | undefined,
): number | null {
  if (value === null || value === undefined || value === "") return null;
  const n = typeof value === "number" ? value : Number(value);
  return Number.isFinite(n) ? n : null;
}

function asInt(
  value: number | string | null | undefined,
  fallback: number | null = 0,
): number | null {
  if (value === null || value === undefined || value === "") return fallback;
  const n = typeof value === "number" ? value : Number(value);
  if (!Number.isFinite(n)) return fallback;
  return Math.trunc(n);
}

export interface B14ReassessmentItemDto {
  id: string;
  improvement_assessment_item_id: string;
  question_version_id: string;
  curriculum_node_id: string;
  item_code_snapshot: string;
  template_kind_snapshot: string;
  question_template_ref_snapshot?: string | null;
}

export interface B14MasteryDeltaDto {
  curriculum_node_id: string;
  baseline_concept_mastery?: number | string | null;
  baseline_execution_accuracy?: number | string | null;
  baseline_concept_decisive_count?: number;
  baseline_execution_decisive_count?: number;
  baseline_concept_inconclusive_count?: number;
  baseline_execution_inconclusive_count?: number;
  baseline_evidence_count?: number;
  baseline_source_evidence_hash?: string;
  post_snapshot_id?: string | null;
  post_published_result_id?: string | null;
  post_concept_mastery?: number | string | null;
  post_execution_accuracy?: number | string | null;
  post_concept_decisive_count?: number | null;
  post_execution_decisive_count?: number | null;
  post_concept_inconclusive_count?: number | null;
  post_execution_inconclusive_count?: number | null;
  post_evidence_count?: number | null;
  post_source_evidence_hash?: string | null;
  concept_delta?: number | string | null;
  execution_delta?: number | string | null;
  materialized_at?: string | null;
  algorithm_version?: string;
}

export interface B14ReassessmentDto {
  id: string;
  improvement_assessment_id: string;
  blueprint_title?: string;
  blueprint_version_number?: number;
  student_id: string;
  curriculum_id: string;
  assessment_id: string;
  assessment_status?: string;
  assessment_version_id: string;
  submission_id?: string | null;
  published_result_id?: string | null;
  status: string;
  algorithm_version?: string;
  instantiation_hash?: string;
  baseline_captured_at?: string;
  created_at?: string;
  updated_at?: string;
  items?: B14ReassessmentItemDto[];
  mastery_deltas?: B14MasteryDeltaDto[];
}

export interface B14RebuildResultDto {
  reassessment_id: string;
  algorithm_version?: string;
  delta_count?: number;
  published_result_id?: string | null;
  source?: string;
}

export function reassessmentItemApiToView(
  dto: B14ReassessmentItemDto,
): ReassessmentItem {
  return {
    id: dto.id,
    improvement_assessment_item_id: dto.improvement_assessment_item_id,
    question_version_id: dto.question_version_id,
    curriculum_node_id: dto.curriculum_node_id,
    item_code_snapshot: dto.item_code_snapshot,
    template_kind_snapshot: dto.template_kind_snapshot,
    question_template_ref_snapshot: dto.question_template_ref_snapshot ?? null,
  };
}

export function reassessmentMasteryDeltaApiToView(
  dto: B14MasteryDeltaDto,
): ReassessmentMasteryDelta {
  return {
    curriculum_node_id: dto.curriculum_node_id,
    baseline_concept_mastery: asNumber(dto.baseline_concept_mastery),
    baseline_execution_accuracy: asNumber(dto.baseline_execution_accuracy),
    baseline_concept_decisive_count:
      asInt(dto.baseline_concept_decisive_count, 0) ?? 0,
    baseline_execution_decisive_count:
      asInt(dto.baseline_execution_decisive_count, 0) ?? 0,
    baseline_concept_inconclusive_count:
      asInt(dto.baseline_concept_inconclusive_count, 0) ?? 0,
    baseline_execution_inconclusive_count:
      asInt(dto.baseline_execution_inconclusive_count, 0) ?? 0,
    baseline_evidence_count: asInt(dto.baseline_evidence_count, 0) ?? 0,
    baseline_source_evidence_hash: dto.baseline_source_evidence_hash ?? "",
    post_snapshot_id: dto.post_snapshot_id ?? null,
    post_published_result_id: dto.post_published_result_id ?? null,
    post_concept_mastery: asNumber(dto.post_concept_mastery),
    post_execution_accuracy: asNumber(dto.post_execution_accuracy),
    post_concept_decisive_count: asInt(dto.post_concept_decisive_count, null),
    post_execution_decisive_count: asInt(
      dto.post_execution_decisive_count,
      null,
    ),
    post_concept_inconclusive_count: asInt(
      dto.post_concept_inconclusive_count,
      null,
    ),
    post_execution_inconclusive_count: asInt(
      dto.post_execution_inconclusive_count,
      null,
    ),
    post_evidence_count: asInt(dto.post_evidence_count, null),
    post_source_evidence_hash: dto.post_source_evidence_hash ?? null,
    concept_delta: asNumber(dto.concept_delta),
    execution_delta: asNumber(dto.execution_delta),
    materialized_at: dto.materialized_at ?? null,
    algorithm_version: dto.algorithm_version ?? "B14_V1",
  };
}

export function reassessmentApiToView(dto: B14ReassessmentDto): Reassessment {
  return {
    id: dto.id,
    improvement_assessment_id: dto.improvement_assessment_id,
    blueprint_title: dto.blueprint_title ?? "",
    blueprint_version_number: dto.blueprint_version_number ?? 1,
    student_id: dto.student_id,
    curriculum_id: dto.curriculum_id,
    assessment_id: dto.assessment_id,
    assessment_status: dto.assessment_status ?? "DRAFT",
    assessment_version_id: dto.assessment_version_id,
    submission_id: dto.submission_id ?? null,
    published_result_id: dto.published_result_id ?? null,
    status: dto.status,
    algorithm_version: dto.algorithm_version ?? "B14_V1",
    instantiation_hash: dto.instantiation_hash ?? "",
    baseline_captured_at:
      dto.baseline_captured_at ?? dto.created_at ?? new Date().toISOString(),
    created_at: dto.created_at ?? new Date().toISOString(),
    updated_at: dto.updated_at ?? dto.created_at ?? new Date().toISOString(),
    items: (dto.items ?? []).map(reassessmentItemApiToView),
    mastery_deltas: (dto.mastery_deltas ?? []).map(
      reassessmentMasteryDeltaApiToView,
    ),
  };
}

export function b14RebuildResultApiToView(
  dto: B14RebuildResultDto,
): B14RebuildResult {
  return {
    reassessment_id: dto.reassessment_id,
    algorithm_version: dto.algorithm_version ?? "B14_V1",
    delta_count: dto.delta_count ?? 0,
    published_result_id: dto.published_result_id ?? null,
    source: dto.source ?? "MASTERY_STATE_SNAPSHOT",
  };
}

/** Live B14 reassessment HTTP adapter. */
export const ReassessmentHttpApi = {
  async instantiateReassessment(
    blueprintId: string,
    input: ReassessmentInstantiateRequest,
  ): Promise<Reassessment> {
    const dto = await httpRequest<B14ReassessmentDto>(
      `/api/v1/improvement-assessments/${blueprintId}/reassessment`,
      { method: "POST", body: input },
    );
    return reassessmentApiToView(dto);
  },

  async getReassessment(id: string): Promise<Reassessment> {
    const dto = await httpRequest<B14ReassessmentDto>(
      `/api/v1/reassessments/${id}`,
    );
    return reassessmentApiToView(dto);
  },

  async rebuildReassessmentB14(id: string): Promise<B14RebuildResult> {
    const dto = await httpRequest<B14RebuildResultDto>(
      `/api/v1/reassessments/${id}/b14/rebuild`,
      { method: "POST" },
    );
    return b14RebuildResultApiToView(dto);
  },
};
